### PR TITLE
[DeviceMesh] Adding nD slicing support back

### LINKED ISSUE #126530: Do we need an N-dim sub-DeviceMesh?
### 🚀 The feature, motivation and pitch

Hey there! Currently `torch.distributed._tensor.DeviceMesh` only supports 1-D sub-meshes, will it be possible to manipulate it as an NDArray and generate N-dim sub-meshes?

For example, in 2-D tensor parallelism together with pipeline parallelism, the mesh looks like `[pp, tp0, tp1]` == `[2, 2, 2]`, and if an all-gather/all-reduce on tp is needed for `rank == 0`, `mesh[tp0]`, `mesh[tp1]` only gives `[0, 1]` and `[0, 2]` (3 is missing).

```python
import os
import torch
from torch.distributed._tensor import DeviceMesh, mesh_resources
rank = int(os.environ['RANK'])
if __name__ == "__main__":
    mesh = DeviceMesh("cpu", [
        [
            [0, 1],
            [2, 3]
        ], # pp_rank == 0
        [
            [4, 5],
            [6, 7]
        ]  # pp_rank == 1
    ], mesh_dim_names=['pp', 'tp0', 'tp1'])
    if rank == 0: # pp_rank == tp0_rank == tp1_rank == 0
        print(mesh['tp0'])
        print(mesh['tp1'])
```

A workaround is to represent `tp0` and `tp1` in a single dim, but there could be scenario in which `tp0` is replicated and only `tp2` needs the communication.

Alternatively, could you please provide some workaround for this?
### Alternatives

_No response_

### Additional context

_No response_

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @msaroufim

### PR DESCRIPTION
**Pull request by wz337**
Fixes #126530

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @tianyu-l @wconstab @yf225 @chauhang @d4l3k @msaroufim

### CODE REVIEW

**Review by wconstab — APPROVED** [review]
lgtm.

**Comment by wz337** [comment]
@huydhn Hmm. I wonder how it is not caught in the first place by the CI. I can add a mypy ignore, but this is what we have today and it looks like it is not complaining. 
https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L92

File: torch/distributed/device_mesh.py [hunk 1]
```diff
@@ -69,31 +69,47 @@ def get_current_mesh(self) -> "DeviceMesh":
...
+            # pg_ranks_by_dim is the size of [number of local ranks of the outermost submesh dimension, *sub_mesh_dims]
+            # This means on each local rank of the outermost slice mesh dim, we have a tensor of submesh size with
+            # the pg ranks of the submesh. From this, we can extract the submesh mesh tensor contains the current rank.
+            pg_ranks_by_dim = parent_mesh.mesh.permute(
+                *mesh_dims_remained, *submesh_dims
+            ).reshape(-1, *submesh_dim_sizes)
+
+            cur_rank = parent_mesh.get_rank()
+            for mesh_nd in pg_ranks_by_dim:
+                # Every rank needs to participate in this DeviceMesh creation even if the cur_rank is not in mesh_nd
```
Comment (wanchaol) [inline torch/distributed/device_mesh.py:96]: Is this comment still true now since we have `init_backend=False`? We don't create any new_groups now during submesh slicing, I think we can probably just create the submesh once for the case when `cur_rank` in mesh_nd

File: torch/distributed/device_mesh.py [hunk 2]
```diff
@@ -69,31 +69,47 @@ def get_current_mesh(self) -> "DeviceMesh":
...
+            mesh_dims_remained = list(range(parent_mesh.mesh.ndim))
+            for submesh_dim in submesh_dims:
+                mesh_dims_remained.remove(submesh_dim)
+
+            # pg_ranks_by_dim is the size of [number of local ranks of the outermost submesh dimension, *sub_mesh_dims]
+            # This means on each local rank of the outermost slice mesh dim, we have a tensor of submesh size with
+            # the pg ranks of the submesh. From this, we can extract the submesh mesh tensor contains the current rank.
+            pg_ranks_by_dim = parent_mesh.mesh.permute(
+                *mesh_dims_remained, *submesh_dims
+            ).reshape(-1, *submesh_dim_sizes)
```
Comment (wanchaol) [inline torch/distributed/device_mesh.py:92]: This is getting a bit complicated, not sure if there's a simpler torch operation that could achieve what we want here 🤔

**Review by wanchaol — APPROVED** [review]
lgtm, thanks for adding this back!

File: torch/distributed/device_mesh.py (hunk 1 above)
Comment (wz337) [reply torch/distributed/device_mesh.py:96]: Oh ye. You are right! Removing this line.

File: torch/distributed/device_mesh.py [hunk 3]
```diff
@@ -69,31 +69,46 @@ def get_current_mesh(self) -> "DeviceMesh":
...
+                    mesh_nd,
+                    mesh_dim_names=submesh_dim_names,
                     _init_backend=False,
                 )
-                if cur_rank in mesh_1d:
-                    res_sub_mesh = sub_mesh
+                if cur_rank in mesh_nd:
+                    res_submesh = submesh
+
+            res_submesh._parent_mesh = parent_mesh  # type: ignore[possibly-undefined]
```
Comment (wconstab) [inline torch/distributed/device_mesh.py:105]: this type: ignore seems suspect to me.  It is not ok if this is undefined.  We would need to ensure it is defined, and typing should help us be honest about that?

File: torch/distributed/device_mesh.py [hunk 4]
```diff
@@ -367,14 +382,14 @@ def __eq__(self, other: object) -> bool:
                     and self._thread_id == other._thread_id
                 )
 
-        def __getitem__(self, mesh_dim_name: str) -> "DeviceMesh":
+        def __getitem__(self, mesh_dim_names: Union[str, Tuple[str]]) -> "DeviceMesh":
```
Comment (wconstab) [inline torch/distributed/device_mesh.py:385]: this typing is incorrect.  It should be `Tuple[str, ...]` to allow for variable length tuples of strings.  Currently, it will only pass linter if called with single dim tuple.

File: torch/distributed/device_mesh.py (hunk 2 above)
Comment (wz337) [reply torch/distributed/device_mesh.py:92]: Not aware of a simpler torch operation. Landing to unblock but can look into this.
